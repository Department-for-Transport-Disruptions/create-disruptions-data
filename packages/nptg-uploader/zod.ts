import { z } from "zod";

const makeArray = <T>(item: T | T[]) => (Array.isArray(item) ? item : [item]);

const adminAreaSchema = z
    .object({
        AdministrativeAreaCode: z.string(),
        AtcoAreaCode: z.string(),
        Name: z.string(),
        ShortName: z.string(),
    })
    .transform((item) => ({
        administrativeAreaCode: item.AdministrativeAreaCode,
        atcoAreaCode: item.AtcoAreaCode,
        name: item.Name,
        shortName: item.ShortName,
    }));

const regionSchema = z.object({
    AdministrativeAreas: z.object({
        AdministrativeArea: z.union([z.array(adminAreaSchema), adminAreaSchema]).transform(makeArray),
    }),
});

const hasAttr = (item: { $: Record<string, unknown> } | string): item is { $: Record<string, unknown> } =>
    !!(item as { $: Record<string, unknown> }).$;

const localitiesQualifySchema = z.object({
    QualifierName: z
        .union([
            z.object({
                $: z.object({
                    "xml:lang": z.string(),
                }),
                _: z.string(),
            }),
            z.string(),
        ])
        .transform((item) => ({
            $: {
                "xml:lang": hasAttr(item) ? item.$["xml:lang"] : null,
            },
            _: hasAttr(item) ? item._ : item,
        }))
        .optional(),
});

const localitiesDescriptorSchema = z.object({
    LocalityName: z.object({
        $: z.object({
            "xml:lang": z.string(),
        }),
        _: z.string(),
    }),
    ShortName: z.string().optional(),
    Qualify: localitiesQualifySchema.optional(),
});

const localitiesSchema = z.object({
    $: z.object({
        CreationDateTime: z.string(),
        ModificationDateTime: z.string(),
        RevisionNumber: z.coerce.number(),
        Modification: z.string(),
    }),
    NptgLocalityCode: z.string(),
    Descriptor: localitiesDescriptorSchema,
    ParentNptgLocalityRef: z
        .object({
            _: z.string(),
        })
        .optional(),
    AdministrativeAreaRef: z.string(),
    NptgDistrictRef: z.string(),
    SourceLocalityType: z.string(),
    Location: z.object({
        Translation: z.object({
            Easting: z.coerce.number(),
            Northing: z.coerce.number(),
        }),
    }),
});

export const nptgSchema = z
    .object({
        NationalPublicTransportGazetteer: z.object({
            Regions: z.object({
                Region: z.array(regionSchema),
            }),
            NptgLocalities: z.object({
                NptgLocality: z.array(localitiesSchema),
            }),
        }),
    })
    .transform((item) => ({
        adminAreas: item.NationalPublicTransportGazetteer.Regions.Region.flatMap(
            (region) => region.AdministrativeAreas.AdministrativeArea,
        ),
        localities: item.NationalPublicTransportGazetteer.NptgLocalities.NptgLocality.flatMap((locality) => ({
            nptgLocalityCode: locality.NptgLocalityCode,
            localityName: locality.Descriptor.LocalityName._,
            localityNameLang: locality.Descriptor.LocalityName.$["xml:lang"],
            shortName: locality.Descriptor.ShortName ?? null,
            shortNameLang: null,
            qualifierName: locality.Descriptor.Qualify?.QualifierName?._ ?? null,
            qualifierNameLang: locality.Descriptor.Qualify?.QualifierName?.$["xml:lang"] ?? null,
            qualifierLocalityRef: null,
            qualifierDistrictRef: null,
            parentLocalityRef: locality.ParentNptgLocalityRef?._,
            administrativeAreaCode: locality.AdministrativeAreaRef,
            nptgDistrictCode: locality.NptgDistrictRef,
            sourceLocalityType: locality.SourceLocalityType,
            gridType: null,
            easting: locality.Location.Translation.Easting,
            northing: locality.Location.Translation.Northing,
            creationDateTime: locality.$.CreationDateTime,
            modificationDateTime: locality.$.ModificationDateTime,
            revisionNumber: locality.$.RevisionNumber,
            modification: locality.$.Modification,
        })),
    }));

export type Nptg = z.infer<typeof nptgSchema>;
