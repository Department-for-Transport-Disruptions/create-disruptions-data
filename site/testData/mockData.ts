/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { mockRequest, mockResponse } from "mock-req-res";
import { NextPageContext } from "next";
import React from "react";
import { Mock, vi } from "vitest";
import { COOKIE_ID_TOKEN, COOKIES_POLICY_COOKIE } from "../constants";

export interface GetMockContextInput {
    session?: { [key: string]: any };
    cookies?: any;
    body?: any;
    url?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    isLoggedin?: boolean;
    query?: any;
}

export const getMockRequestAndResponse = ({
    cookieValues = {},
    body = null,
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    requestHeaders = {},
    isLoggedin = true,
    url = null,
    query = null,
}: GetMockRequestAndResponse = {}): { req: any; res: any } => {
    const res = mockResponse({ writeHead: mockWriteHeadFn, end: mockEndFn });

    const {
        idToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJjdXN0b206bm9jIjoiVEVTVCIsImVtYWlsIjoidGVzdEBleGFtcGxlLmNvbSIsImp0aSI6Ijg1MmQ1MTVlLTU5YWUtNDllZi1iMTA5LTI4YTRhNzk3YWFkNSIsImlhdCI6MTU5Mjk4NzMwNywiZXhwIjoxNTkyOTkwOTA3fQ.DFdxnpdhykDONOMeZMNeMUFpCHZ-hQ3UXczq_Qh0IAI",
        cookiePolicy = null,
    } = cookieValues;

    const defaultSession = {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        destroy: (): void => {},
    };

    let cookieString = "";

    cookieString += isLoggedin ? `${COOKIE_ID_TOKEN}=${idToken as string};` : "";

    cookieString += cookiePolicy ? `${COOKIES_POLICY_COOKIE}=${encodeURI(JSON.stringify(cookiePolicy))}` : "";

    const req = mockRequest({
        connection: {
            encrypted: true,
        },
        url,
        headers: {
            host: "localhost:3000",
            cookie: cookieString,
            origin: "localhost:3000",
            ...requestHeaders,
        },
        cookies: cookieValues,
        session: { ...defaultSession },
    });

    if (body) {
        req.body = body;
    }
    if (query) {
        req.query = query;
    }

    return { req, res };
};

export const getMockContext = ({
    cookies = {},
    body = null,
    uuid = {},
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    isLoggedin = true,
    url = null,
    query = "",
}: GetMockContextInput = {}): NextPageContext => {
    const { req, res } = getMockRequestAndResponse({
        cookieValues: cookies,
        body,
        uuid,
        mockWriteHeadFn,
        mockEndFn,
        requestHeaders: {},
        isLoggedin,
        url,
    });

    const ctx: NextPageContext = {
        res,
        req,
        pathname: "",
        query,
        // eslint-disable-next-line react/display-name
        AppTree: () => React.createElement("div"),
    };

    return ctx;
};

export interface GetMockRequestAndResponse {
    cookieValues?: any;
    body?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    requestHeaders?: any;
    isLoggedin?: boolean;
    url?: any;
    query?: any;
}
