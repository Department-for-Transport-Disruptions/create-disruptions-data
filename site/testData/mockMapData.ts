export const mockMarkerData = [
    {
        atcoCode: "1800SB12581",
        commonName: "Cheapside",
        indicator: "Stop CL",
        longitude: -2.244551,
        latitude: 53.48096,
        bearing: "W",
        stopType: "BCT",
        busStopType: "MKD",
    },
    {
        atcoCode: "1800SB05801",
        commonName: "Cheapside",
        indicator: "Stop CO",
        longitude: -2.243361,
        latitude: 53.48097,
        bearing: "E",
        stopType: "BCT",
        busStopType: "MKD",
    },
    {
        atcoCode: "1800SB05751",
        commonName: "Manchester Town Hall",
        indicator: "Stop SC",
        longitude: -2.244697,
        latitude: 53.480024,
        bearing: "SE",
        stopType: "BCT",
        busStopType: "MKD",
    },
];
