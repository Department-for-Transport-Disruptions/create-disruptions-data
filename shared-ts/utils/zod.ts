import { ZodErrorMap, ZodSchema, z } from "zod";

export const setZodDefaultError: (errorMessage: string) => { errorMap: ZodErrorMap } = (errorMessage: string) => ({
    errorMap: (issue) => {
        switch (issue.code) {
            default:
                return { message: errorMessage };
        }
    },
});

const dateRegex = /^(0?[1-9]|[12][0-9]|3[01])[\/](0?[1-9]|1[012])[\/]\d{4}$/;
const timeRegex = /^([0-1][0-9]|2[0-3])[0-5][0-9]$/;
const minutesRegex = /^[0-9]{0,3}$/;
const uuidRegex = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/i;

/**
 * Verify if the input value is of date format DD/MM/YYYY.
 * @param {defaultError} defaultError Error message when the validation fails.
 *
 * @returns {z.ZodString} Indicates an error if the input format is not DD/MM/YYYY.
 */

export const zodDate = (defaultError?: string) =>
    z.string(defaultError ? setZodDefaultError(defaultError) : {}).regex(dateRegex);

/**
 * Verify if the input value is of format hhmm where hh is 0-23 and mm is 0-59.
 * @param {defaultError} defaultError Error message when the validation fails.
 *
 * @returns {z.ZodString} Indicates an error if the regex match fails and the value
 *  is not of format hhmm.
 */
export const zodTime = (defaultError?: string) =>
    z.string(defaultError ? setZodDefaultError(defaultError) : {}).regex(timeRegex);

/**
 * Verify if the input value is a number between 0-999 minutes.
 * @param {defaultError} defaultError Error message when the validation fails.
 *
 * @returns {z.ZodString} Indicates an error if the regex match fails.
 */
export const zodTimeInMinutes = (defaultError?: string) =>
    z.string(defaultError ? setZodDefaultError(defaultError) : {}).regex(minutesRegex);

/**
 * Verify if the input value matches uuid regex expression
 * This is because the current version of Zod (3.21.4) does not support uuid version 7
 * @param {defaultError} defaultError Error message when the validation fails.
 *
 * @returns {z.ZodString} Indicates an error if the input format is the regex match fails.
 */
export const zodUuid = (defaultError?: string) =>
    z.string(defaultError ? setZodDefaultError(defaultError) : {}).regex(uuidRegex);

/**
 * Returns a zod schema that can be used to filter out items that do not conform to the
 * given schema
 * @param {ZodSchema} schema Zod schema to use
 *
 * @returns {ZodSchema} Zod schema that can be used to filter out invalid items
 */
export const makeFilteredArraySchema = <T extends ZodSchema>(schema: T) =>
    z
        .array(z.unknown())
        .transform((items) => items?.filter((item): item is z.infer<T> => schema.safeParse(item).success));
