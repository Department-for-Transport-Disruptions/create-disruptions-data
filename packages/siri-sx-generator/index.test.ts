import { S3Client, PutObjectCommand } from "@aws-sdk/client-s3";
import { DynamoDBDocumentClient, ScanCommand } from "@aws-sdk/lib-dynamodb";
import { mockClient } from "aws-sdk-client-mock";
import { describe, expect, it, beforeEach, beforeAll } from "vitest";
import formatXml from "xml-formatter";
import { ZodError } from "zod";
import { invalidDisruptionJsonExamples, testDisruptionsJson } from "./test/testData";
import { generateSiriSxAndUploadToS3 } from ".";

const ddbMock = mockClient(DynamoDBDocumentClient);
const s3Mock = mockClient(S3Client);

describe("SIRI-SX Generator", () => {
    beforeAll(() => {
        process.env.TABLE_NAME = "test-table";
        process.env.SIRI_SX_UNVALIDATED_BUCKET_NAME = "test-bucket";
    });

    beforeEach(() => {
        ddbMock.reset();
        s3Mock.reset();
    });

    it("correctly generates SIRI-SX XML", async () => {
        ddbMock.on(ScanCommand).resolves({ Items: testDisruptionsJson });

        await generateSiriSxAndUploadToS3(
            s3Mock as unknown as S3Client,
            ddbMock as unknown as DynamoDBDocumentClient,
            "test-table",
            "test-bucket",
            "abcde-fghij-klmno-pqrst",
            "2023-03-06T12:00:00Z",
        );

        const s3PutCommand = s3Mock.commandCalls(PutObjectCommand)[0].args[0];
        const putData = (s3PutCommand.input.Body as string).replace(/(?:\r\n|\r|\n)/g, "");

        expect(s3PutCommand.input.Key).toBe("1678104000000-unvalidated-siri.xml");
        expect(
            formatXml(putData, {
                collapseContent: true,
            }),
        ).toMatchSnapshot();
    });

    it.each(invalidDisruptionJsonExamples)("handles invalid disruptions JSON - %s", async (_description, input) => {
        ddbMock.on(ScanCommand).resolves({ Items: [input] });

        await expect(
            generateSiriSxAndUploadToS3(
                s3Mock as unknown as S3Client,
                ddbMock as unknown as DynamoDBDocumentClient,
                "test-table",
                "test-bucket",
                "abcde-fghij-klmno-pqrst",
                "2023-03-06T12:00:00Z",
            ),
        ).rejects.toThrowError(ZodError);

        await expect(
            generateSiriSxAndUploadToS3(
                s3Mock as unknown as S3Client,
                ddbMock as unknown as DynamoDBDocumentClient,
                "test-table",
                "test-bucket",
                "abcde-fghij-klmno-pqrst",
                "2023-03-06T12:00:00Z",
            ),
        ).rejects.toThrowErrorMatchingSnapshot();
    });
});
