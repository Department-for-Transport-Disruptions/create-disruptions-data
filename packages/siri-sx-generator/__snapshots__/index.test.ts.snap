// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SIRI-SX Generator > correctly generates Disruptions CSV 1`] = `
"Organisation,ID,Validity start,Validity end,Publication start,Publication end,Reason,Planned,Modes affected,Operators affected,Services affected,Stops affected
Test Org,acde070d-8c4c-4f0d-9d8a-162843c10333,2023-03-10T12:00:00.000Z,,2023-03-10T12:00:00.000Z,,staffInWrongPlace,true,bus;tram,FSYO,,"
`;

exports[`SIRI-SX Generator > correctly generates Disruptions JSON 1`] = `
[
  {
    "consequences": [
      {
        "consequenceOperators": [
          {
            "operatorNoc": "FSYO",
            "operatorPublicName": "Operator Name",
          },
        ],
        "consequenceType": "operatorWide",
        "description": "Some consequence description",
        "disruptionDelay": "40",
        "disruptionSeverity": "severe",
        "removeFromJourneyPlanners": "yes",
        "vehicleMode": "bus",
      },
      {
        "consequenceType": "networkWide",
        "description": "Some consequence description",
        "disruptionSeverity": "slight",
        "removeFromJourneyPlanners": "no",
        "vehicleMode": "tram",
      },
    ],
    "description": "Test description",
    "displayId": "8fg3ha",
    "disruptionId": "acde070d-8c4c-4f0d-9d8a-162843c10333",
    "disruptionReason": "staffInWrongPlace",
    "disruptionType": "planned",
    "history": [
      {
        "datetime": "2023-05-19T14:40:00Z",
        "historyItems": [
          "some text",
        ],
        "status": "PENDING_APPROVAL",
        "user": "user",
      },
      {
        "datetime": "2023-05-13T14:45:00Z",
        "historyItems": [
          "Disruption created and published",
        ],
        "status": "PUBLISHED",
        "user": "user",
      },
    ],
    "lastUpdated": "2023-05-11T14:40:00Z",
    "organisation": {
      "id": "6f6435e3-a485-4c8c-8c29-e121b1e76802",
      "name": "Test Org",
    },
    "publishStartDate": "10/03/2023",
    "publishStartTime": "1200",
    "summary": "Some summary",
    "validity": [
      {
        "disruptionEndDate": "17/03/2023",
        "disruptionEndTime": "1700",
        "disruptionStartDate": "10/03/2023",
        "disruptionStartTime": "1200",
      },
      {
        "disruptionStartDate": "18/03/2023",
        "disruptionStartTime": "1200",
      },
    ],
  },
]
`;

exports[`SIRI-SX Generator > correctly generates SIRI-SX XML 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?>
<Siri version=\\"2.0\\" xmlns=\\"http://www.siri.org.uk/siri\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.siri.org.uk/siri http://www.siri.org.uk/schema/2.0/xsd/siri.xsd\\">
    <ServiceDelivery>
        <ResponseTimestamp>2023-08-17T00:00:00Z</ResponseTimestamp>
        <ProducerRef>DepartmentForTransport</ProducerRef>
        <ResponseMessageIdentifier>abcde-fghij-klmno-pqrst</ResponseMessageIdentifier>
        <SituationExchangeDelivery>
            <ResponseTimestamp>2023-08-17T00:00:00Z</ResponseTimestamp>
            <Situations>
                <PtSituationElement>
                    <CreationTime>2023-05-13T14:45:00Z</CreationTime>
                    <VersionedAtTime>2023-05-11T14:40:00Z</VersionedAtTime>
                    <ParticipantRef>TestOrg</ParticipantRef>
                    <SituationNumber>acde070d-8c4c-4f0d-9d8a-162843c10333</SituationNumber>
                    <Source>
                        <SourceType>feed</SourceType>
                        <TimeOfCommunication>2023-08-17T00:00:00.000Z</TimeOfCommunication>
                    </Source>
                    <Progress>open</Progress>
                    <ValidityPeriod>
                        <StartTime>2023-03-10T12:00:00.000Z</StartTime>
                        <EndTime>2023-03-17T17:00:00.000Z</EndTime>
                    </ValidityPeriod>
                    <ValidityPeriod>
                        <StartTime>2023-03-18T12:00:00.000Z</StartTime>
                    </ValidityPeriod>
                    <PublicationWindow>
                        <StartTime>2023-03-10T12:00:00.000Z</StartTime>
                    </PublicationWindow>
                    <PersonnelReason>staffInWrongPlace</PersonnelReason>
                    <Planned>true</Planned>
                    <Summary>Some summary</Summary>
                    <Description>Test description</Description>
                    <Consequences>
                        <Consequence>
                            <Condition>unknown</Condition>
                            <Severity>severe</Severity>
                            <Affects>
                                <Operators>
                                    <AffectedOperator>
                                        <OperatorRef>FSYO</OperatorRef>
                                        <OperatorName>Operator Name</OperatorName>
                                    </AffectedOperator>
                                </Operators>
                                <Networks>
                                    <AffectedNetwork>
                                        <VehicleMode>bus</VehicleMode>
                                        <AllLines/>
                                    </AffectedNetwork>
                                </Networks>
                            </Affects>
                            <Advice>
                                <Details>Some consequence description</Details>
                            </Advice>
                            <Blocking>
                                <JourneyPlanner>true</JourneyPlanner>
                            </Blocking>
                            <Delays>
                                <Delay>PT40M</Delay>
                            </Delays>
                        </Consequence>
                        <Consequence>
                            <Condition>unknown</Condition>
                            <Severity>slight</Severity>
                            <Affects>
                                <Operators>
                                    <AllOperators/>
                                </Operators>
                                <Networks>
                                    <AffectedNetwork>
                                        <VehicleMode>tram</VehicleMode>
                                        <AllLines/>
                                    </AffectedNetwork>
                                </Networks>
                            </Affects>
                            <Advice>
                                <Details>Some consequence description</Details>
                            </Advice>
                            <Blocking>
                                <JourneyPlanner>false</JourneyPlanner>
                            </Blocking>
                        </Consequence>
                    </Consequences>
                </PtSituationElement>
            </Situations>
        </SituationExchangeDelivery>
    </ServiceDelivery>
</Siri>"
`;

exports[`SIRI-SX Generator > correctly generates SIRI-SX XML where creationDate is present 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?>
<Siri version=\\"2.0\\" xmlns=\\"http://www.siri.org.uk/siri\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.siri.org.uk/siri http://www.siri.org.uk/schema/2.0/xsd/siri.xsd\\">
    <ServiceDelivery>
        <ResponseTimestamp>2023-03-06T12:00:00Z</ResponseTimestamp>
        <ProducerRef>DepartmentForTransport</ProducerRef>
        <ResponseMessageIdentifier>abcde-fghij-klmno-pqrst</ResponseMessageIdentifier>
        <SituationExchangeDelivery>
            <ResponseTimestamp>2023-03-06T12:00:00Z</ResponseTimestamp>
            <Situations>
                <PtSituationElement>
                    <CreationTime>2023-05-13T14:45:00Z</CreationTime>
                    <VersionedAtTime>2023-05-11T14:40:00Z</VersionedAtTime>
                    <ParticipantRef>TestOrg</ParticipantRef>
                    <SituationNumber>acde070d-8c4c-4f0d-9d8a-162843c10333</SituationNumber>
                    <Source>
                        <SourceType>feed</SourceType>
                        <TimeOfCommunication>2023-08-17T00:00:00.000Z</TimeOfCommunication>
                    </Source>
                    <Progress>open</Progress>
                    <ValidityPeriod>
                        <StartTime>2023-03-10T12:00:00.000Z</StartTime>
                        <EndTime>2023-03-17T17:00:00.000Z</EndTime>
                    </ValidityPeriod>
                    <ValidityPeriod>
                        <StartTime>2023-03-18T12:00:00.000Z</StartTime>
                    </ValidityPeriod>
                    <PublicationWindow>
                        <StartTime>2023-03-10T12:00:00.000Z</StartTime>
                    </PublicationWindow>
                    <PersonnelReason>staffInWrongPlace</PersonnelReason>
                    <Planned>true</Planned>
                    <Summary>Some summary</Summary>
                    <Description>Test description</Description>
                    <Consequences>
                        <Consequence>
                            <Condition>unknown</Condition>
                            <Severity>severe</Severity>
                            <Affects>
                                <Operators>
                                    <AffectedOperator>
                                        <OperatorRef>FSYO</OperatorRef>
                                        <OperatorName>Operator Name</OperatorName>
                                    </AffectedOperator>
                                </Operators>
                                <Networks>
                                    <AffectedNetwork>
                                        <VehicleMode>bus</VehicleMode>
                                        <AllLines/>
                                    </AffectedNetwork>
                                </Networks>
                            </Affects>
                            <Advice>
                                <Details>Some consequence description</Details>
                            </Advice>
                            <Blocking>
                                <JourneyPlanner>true</JourneyPlanner>
                            </Blocking>
                            <Delays>
                                <Delay>PT40M</Delay>
                            </Delays>
                        </Consequence>
                        <Consequence>
                            <Condition>unknown</Condition>
                            <Severity>slight</Severity>
                            <Affects>
                                <Operators>
                                    <AllOperators/>
                                </Operators>
                                <Networks>
                                    <AffectedNetwork>
                                        <VehicleMode>tram</VehicleMode>
                                        <AllLines/>
                                    </AffectedNetwork>
                                </Networks>
                            </Affects>
                            <Advice>
                                <Details>Some consequence description</Details>
                            </Advice>
                            <Blocking>
                                <JourneyPlanner>false</JourneyPlanner>
                            </Blocking>
                        </Consequence>
                    </Consequences>
                </PtSituationElement>
            </Situations>
        </SituationExchangeDelivery>
    </ServiceDelivery>
</Siri>"
`;
